# Функция

**Функция — это фрагмент программного кода, к которому можно обратиться из другого места программы.** (Следует понимать, что «другое место программы» совсем не обязательно должно находиться в том же самом файле, где написана функция. Например, исходный код встроенных в Python функций находится на компьютере непосредственно в папке с самим Python, и интерпретатор использует функции оттуда.)



**Аргумент (параметр) — это независимая переменная, значение которой используется функцией для выполнения своего исходного кода.** Словосочетание «независимая переменная» означает, что объект, переданный в функцию для обработки, изначально не содержится в функции, а передаётся в неё извне.

#### Функции позволяют:

- выполнять один и тот же набор инструкций (фрагмент исходного кода) несколько раз;
- выполнять одни и те же действия с различными входными данными;
- структурировать исходный код.

In [1]:
# Пример функции:
def first_function():
    print("Hello world!")
first_function()
first_function()

Hello world!
Hello world!


# Сигнатура и сематика функции
- Сигнатура (название функции) - непосредственный способ назвать функцию, отличить её от остальных.
- Семматика - (смысл функции) исходный код, следующий после доветочия до окончания отступа в 4 пробела, задает выполняемые функцией действия, придаёт ей смысл. Однако чаще вам будет встречаться выражение **тело функции**, которое означает то же самое.


Рассмотрим подробно первую строку:

1. Слово def является общим для любой функции — оно даёт интерпретатору понять, что далее следует описание новой функции.

2. Сразу через пробел после def следует название функции, которое может быть любым.

→ Требования к названиям функции такие же, как и к названиям переменных — вы их уже изучали.

→ Важно, чтобы названия функций не повторялись и не использовали названия встроенных функций, таких как print, list, sum и т. д. Если вы создадите функцию с повторяющимся названием, вы просто перезапишете её и уже не сможете воспользоваться исходной версией.

→ Старайтесь, чтобы по названию можно было хотя бы примерно понять, какие действия она выполняет.

3. Сразу после названия функции без пробела следуют круглые скобки. В них перечисляются названия аргументов (требования к названиям аргументов тоже совпадают с требованиями к названиям переменных), которые функция будет принимать.

→ Даже если для данной функции не требуются аргументы (как в примере выше), круглые скобки обязательны!

4. Сразу после закрывающейся круглой скобки следует двоеточие, которое позволяет интерпретатору понять, какие именно действия выполняет созданная вами функция.

5. После двоеточия с новой строки и с отступом в 4 пробела от начала слова def следует исходный код функции.

→ Важно, чтобы отступ в 4 пробела сохранялся до конца исходного кода функции. Как только отступ вернётся на уровень слова def, компьютер поймёт, что исходный код функции окончен

In [4]:
def first_function():
    # первая строка функции
    print("Hello world!")
    # вторая строка функции
    print("Line 2")
# Отступ вернулся на прежний уровень, значит
# код функции закончился.

Как теперь воспользоваться написанной нами функцией?

Необходимо написать в скрипте название функции, не забыв добавить к нему круглые скобки.

Вот так можно запустить созданную нами функцию first_function:

In [5]:
first_function()

Hello world!
Line 2


Ещё пример функции:


In [6]:
def print_hours(minutes):
    # // — это оператор целочисленного деления
    hours = minutes // 60
    # % — это оператор получения остатка от деления
    left_minutes = minutes % 60
    print("Hours:", hours)
    print("Minutes left:", left_minutes)
    
print_hours(90)

Hours: 1
Minutes left: 30


# Оператор Return

- return - оператор, который позволяет функции передать какой-либо объект (число, строку, список, даже другую функцию) в то место, кода, откуда вызывалась фукция.

Пример функции, которая принимает на вход расстояние и среднюю скорость и возвращает время в пути:

In [7]:
# Назовем функцию get_time (get - получать,
# time - время). Она принимает аргументы
# distance - расстояние и speed - скорость.
def get_time(distance, speed):
    result = distance / speed
# Чтобы вернуть результат вычислений,
# пишем оператор return И название переменной,
# значение которой будет передано.
    return result

- функция готова - теперь её можно запустить непосредственно из исходного кода скрипта.
- Передавать аргументы в функцию необходимо в том же самом порядке, в котором они перечисояются в сигнатуре функции: в данном случае первым идёт расстояние, а зател идёт.

In [11]:
time = get_time(120, 60)
print(int(time))

2


# Return и несколько переменныхх

In [13]:
def get_time_tuple(distance, speed):
    hours = distance // speed
    distance_left = distance % speed
    kms_per_minute = speed / 60
    minutes = round (distance_left / kms_per_minute)
    # Аргументы будут возвращены функцией в виде кортежа
    return hours, minutes

result = get_time_tuple(120, 100)
print (result)


(1, 12)


Важно, чтобы количество переменных перед знаком = совпадало с количеством переменных, которое возвращает функция.

In [14]:
# Через запятую перечисляем переменные, в которые сохранится результат
hours, minutes = get_time_tuple(120, 100)
# Красиво напечатаем результат
print("Hours to travel:", hours)
print("Minutes to travel:", minutes)
# Будет напечатано:
# Hours to travel: 1
# Minutes to travel: 12

Hours to travel: 1
Minutes to travel: 12


# Особенности использования return

*Замечание 1* - Если функция печатает что на экран, это не означает, что она возвращает то же значение.

*Замечание 2* - Python автоматиччески возвращает None, если разработчик самостоятельно не прописал, что должна возвращать функция.

*Замечание 3* - Когда компьютер встречает **return** при выполнении кода функции, он автоматически прекращает выполнение функции, даже если за строкой с **return** следуют дополнительные инструкции.

Проверим всё это на примере функции, которая принимает на вход список и объект, который необходимо в нём найти. Она ищет его в цикле и возвращает True, если объект найден, и False, если его в списке нет.

In [16]:
# list_in — список, в котором будем искать объект
# Интуитивно хотелось бы назвать аргумент для списка
# словом list, однако это привело бы к изменению встроенного
# объекта list, что очень нежелательно
# obj — аргумент, содержащий объект, который ищет функция
def in_list(list_in, obj):
    for elem in list_in:
        if obj == elem:
            print("Element is found!")
            return True
            print("This won’t be printed")
    print("Element is NOT found!")
    return False
    print("This will not be printed either")

In [17]:
my_list = [1,2,5,7,10]
result = in_list(my_list, 3)
# Element is NOT found!
print(result)
# False
result = in_list(my_list, 7)
# Element is found!
print(result)
# True

Element is NOT found!
False
Element is found!
True


# ФУНКЦИИ КАК ИНСТРУМЕНТ ПРОЕКТИРОВАНИЯ СТРУКТУРЫ ПРОГРАММЫ

Иногда бывает удобно задать только сигнатуру функции, а конкретно семантику расписать уже в процессе разработки.Тогда необходимо заполнить пустую функцию позволяет слово pass. Оно означает, что выполнять никакое действие не требуется, однако компьютер понимает, что разработчик не забыл про отступ после двоеточия, и ошибка не возникает.

Например, мы хотим разработать программу для управления лифтом. В голову сразу приходят три функции, которые должен уметь выполнять лифт: открывать и закрывать двери, перемещаться на заданный этаж.

Вот так выглядела бы заготовка для программы управления лифтом:


In [18]:
# Функция открытия дверей
def open_doors():
    pass
# Функция закрытия дверей
def close_doors():
    pass
# Функция перемещения на этаж с номером number
def move_to_floor(number):
    pass

- Иногда, чтобы структурировать написание кода, бывает удобно сразу прописать в коде названия функций, которые будут использоваться, а вернуться к написанию их исходного кода немного позднее.

# Проверка аргументов на корректность

Посмотрим, какая ошибка возникает, если попробовать получить значение из словаря по ключу, которого в нём нет:


In [19]:
grades = {'Ivanov': 5, 'Smirnov': 3, 'Kuznetsova': 4, 'Tihonova': 5}
# Напечатаем оценку студента, которого нет в словаре
print(grades['Pavlov'])

KeyError: 'Pavlov'

Исключение KeyError: 'Pavlov': оно означает, что переданного ключа (слово "Pavlov") нет в словаре.

Теперь обработаем это исключение, то есть покажем интерпретатору, что:

- нам известно о том, что такая ошибка может возникнуть;
- мы предусмотрели, что делать в этом случае;
- аварийное прекращение работы скрипта не требуется.

Код будет выглядеть вот так:

In [20]:
# Создадим тот же словарь
grades = {'Ivanov': 5, 'Smirnov': 3, 'Kuznetsova': 4, 'Tihonova': 5}
# Только попробуем (try — пробовать) напечатать оценку студента,
# которого нет в словаре
try:
    print(grades['Pavlov'])
# А если возникнет ошибка в ключе (KeyError), скажем,
# что студента нет в словаре
except KeyError:
    print("Student’s mark was not found!")

Student’s mark was not found!


Разберем ошибки на примере функции:

Возьмем выше написанную функцию. Чтобы случайно не передать в функцию get_time некорректные аргументы и не допустить ошибку, добавим их проверку. Если окажется, что аргументы меньше нуля, вызовем ValueError с помощью оператора raise.

> После встречи с raise, как и в случае с return, интерпретатор прекращает исполнение кода функции.


In [31]:
def get_time(distance, speed):
    # Если расстояние или скорость отрицательные, то возвращаем ошибку
    if distance < 0 or speed < 0:
        # Оператор raise возвращает (raise — досл. англ. "поднимать")
        # объект-исключение. В данном случае ValueError (некорректное значение).
        # Дополнительно в скобках после слова ValueError пишем текст сообщения
        # об ошибке, чтобы сразу было понятно, чем вызвана ошибка.
        raise ValueError("Distance or speed cannot be below 0!")
    result = distance / speed
    return result

In [32]:
get_time(50, -50)

ValueError: Distance or speed cannot be below 0!

Теперь при попытке передать некорректные аргументы вы сразу получаете сообщение об ошибке и понимаете, что входные данные необходимо изменить, чтобы они соответствовали требованиям функции.

# Аргументы по умолчанию

Можно создавать функции, в которые можно передавать не все аргументы, а, например, задать часть аргументов (или даже все) аргументами по умолчанию, то есть сообщить функции, какие данные использовать, если они не были переданы в функцию извне.

> Аргументы по умолчанию задаются в Python очень просто: достаточно после названия переменной сразу же прописать знак = (равно) и то значение, которое вы бы хотели присвоить по умолчанию.

Давайте напишем шуточную функцию get_cola, которая принимает на вход только один аргумент — ice. ice будет принимать значения True или False, причём по умолчанию будет задано True. В зависимости от True или False функция будет печатать "Кола со льдом готова!" или "Кола безо льда готова!", соответственно.


In [35]:
def get_cola (ice=True):
    if ice == True:
        print('Cola with ice is ready!')
    else:
        print('Cola without ice is ready!')
#Теперь запустим жту функцию с разными аргументамиx
print ('С аргументами')
get_cola(True)
get_cola(False)
print()
print ('С аргументами по умолчанию')
get_cola()

С аргументами
Cola with ice is ready!
Cola without ice is ready!

С аргументами по умолчанию
Cola with ice is ready!


- Функция сработала и без передачи аргументов, и мы по умолчанию получили колу со льдом.

Посмотрим теперь, что бы произошло, если бы мы не указали при создании функции значение по умолчанию для аргумента ice:



In [36]:
# Аргументу ice не присваиваем значение по умолчанию:
def get_cola(ice):
    if ice == True:
        print("Cola with ice is ready!")
    else:
        print("Cola without ice is ready!")
 
get_cola()

TypeError: get_cola() missing 1 required positional argument: 'ice'

-  Вы заметили, что возникла ошибка. Она переводится так: «В функции get_cola() не хватает одного обязательного позиционного аргумента 'ice'».

> Итак, использование аргументов по умолчанию позволяет передавать в функцию не все аргументы и не получать из-за этого сообщение об ошибке. Можно сказать, что если у аргумента есть значение по умолчанию, он становится необязательным.

Важное замечание. Указывать аргументам значения по умолчанию можно только после того, как в функции перечислены все обязательные аргументы.

Эта ошибка переводится так: «Синтаксическая ошибка: аргумент без значения по умолчанию следует после аргумента по умолчанию». Чтобы её избежать, необходимо указывать аргументы по умолчанию только после обязательных.

Поменяем в модифицированной функции root порядок аргументов:

In [37]:
def wrong_root(n=2, value):
    result = value ** (1/n)
    return result

SyntaxError: non-default argument follows default argument (3187975254.py, line 1)

# Опасность аргументов по умолчанию

- Использование аргументов по умолчанию призвано упростить жизнь разработчику. Но, к сожалению, их некорректное использование может приводить к крайне неожиданным последтсвиям.

Создадим функцию, поведение которой будет неверным (и заодно повторим использование словарей).

Функция add_mark будет записывать оценку студента в словарь. Словарь не требуется обязательно передавать в функцию: будем использовать пустой словарь в качестве аргумента по умолчанию (⛔ сейчас будет пример того, как делать не нужно!).

Итак, получим функцию:

In [38]:
# Функция add_mark (от англ. add mark — "добавить оценку")
# принимает обязательные аргументы name (имя) и
# mark (оценка). Необязательным аргументом является journal
# (журнал оценок), который по умолчанию является пустым словарём.
def add_mark(name, mark, journal={}):
    # Присваиваем имени в журнале переданную оценку
    journal[name] = mark
    return journal

Теперь будем экспериментировать с новой функцией. Вначале передадим все три аргумента извне:

In [39]:
# Создадим пустой словарь, в который будем
# сохранять оценки группы 1
group1 = {}
# Добавим оценки двух студентов
group1 = add_mark('Ivanov', 5, group1)
group1 = add_mark('Tihonova', 4, group1)
print(group1)

{'Ivanov': 5, 'Tihonova': 4}


Поведение функции выглядит именно так, как мы и задумывали: в новый журнал оценок (тип данных: словарь) добавляем оценки студентов, причём пустой журнал мы создали в основном скрипте, а не использовали его из значения по умолчанию.

А теперь не будем передавать пустой журнал изначально в качестве аргумента — он ведь и так создаётся по умолчанию:

In [40]:
# Добавим в журнал для новой группы оценку Смирнову
# Сам журнал не передаём в виде аргумента
# Пустой журнал будет использован как аргумент по умолчанию
group2 = add_mark('Smirnov', 3)
print(group2)

 
# Аналогично для новой группы 3 добавим оценку Кузнецовой
group3 = add_mark('Kuznetsova', 5)
print(group3)

{'Smirnov': 3}
{'Smirnov': 3, 'Kuznetsova': 5}


Очень неожиданно, не так ли? Мы думали, что по умолчанию используется пустой словарь для группы 3, но в нём почему-то уже оказались оценки для второй группы. Как же так?

Дело в том, что пустой словарь был присвоен как новый объект аргументу journal только один раз: когда интерпретатор впервые прочитал объявление функции в коде. В аргументе journal по умолчанию теперь хранится указатель на созданный при прочтении функции словарь. В первый раз **[group2 = add_mark('Смирнов', 3)]** мы действительно воспользовались новым словарём. А вот при повторном использовании данного словаря в качестве аргумента по умолчанию **[group3 = add_mark('Кузнецова', 5)]**, мы модифицировали уже созданный ранее словарь.

- Если вы не поняли предыдущее объяснение, попробуйте осознать следующую иллюстрацию и сопоставить её с описанной выше проблемой.

Допустим, в учебную часть пришел преподаватель и спросил, куда ставить оценки за экзамен, если в кабинете нет ведомости. Ему сказали, что новая ведомость лежит в третьем ящике стола у входа (то есть преподаватель получил указатель на пустую ведомость, узнал, где она должна храниться).

Преподаватель пришёл на экзамен, ведомости не было, поэтому он взял ведомость из третьего ящика стола (перешёл по указателю), вписал туда оценки (изменил данный объект) и вернул ведомость на место.

Этот же преподаватель пришёл принимать экзамен у ещё одной группы. Ведомость снова никто не принёс, но он помнит, что в таком случае необходимо использовать ведомость из третьего ящика стола (снова переходит по указателю). Преподаватель внёс в эту ведомость новые оценки и вернул её на место.

Что же теперь записано в ведомости? Оценки обеих групп! Преподаватель только один раз получил указания, где брать ведомость, если её нет, и записывал оценки только в неё.

То же самое происходит и с пустым словарём, который мы установили в виде аргумента по умолчанию: он был создан только 1 раз, а затем он только модифицировался функцией, а не создавался пустым заново.

Давайте сделаем так, чтобы каждый раз при запуске функции без словаря всё-таки создавался пустой словарь, а не использовался однажды созданный. Для этого изначально присвоим аргументу journal значение по умолчанию None и уже непосредственно в теле функции будем создавать пустой словарь, если journal является None:

In [42]:
def add_mark(name, mark, journal=None):
    # Если журнал является None
    # (напоминание: сравнивать объект с None
    # корректнее через оператор is),
    # запишем в journal пустой словарь
    if journal is None:
        journal = {}
    journal[name] = mark
    return journal

group2 = add_mark('Smirnov', 3)
print(group2)

 
group3 = add_mark('Kuznetsova', 5)
print(group3)


{'Smirnov': 3}
{'Kuznetsova': 5}


Как видите, теперь всё встало на свои места: оценки для новых групп теперь действительно записываются в новый словарь, а не добавляются в ранее созданный.

- Теперь остаётся понять, какие объекты не рекомендуется использовать в качестве аргументов по умолчанию.

К сожалению, полностью осознать ответ на этот вопрос можно будет только после изучения **объектно-ориентированного программирования**, то есть через несколько модулей.


#### Важно!

- **В качестве аргументов по умолчанию некорректно использовать изменяемые типы данных: списки, словари, множества…** У всех из них есть собственные функции и методы, с помощью которых происходят изменения этих объектов (например, append, add и т. д.) непосредственно внутри объекта.
- **В качестве аргументов по умолчанию точно можно использовать «простые» типы данных**, которые не содержат в себе дополнительные значения, такие как int, float, str, bool, None.

Остальные типы данных, скорее всего, состоят из перечисленных «простых» и потому являются «сложными», например списки могут состоять из всех перечисленных выше «простых» элементов.

Ввиду особенностей работы со «сложными» типами данных, они являются изменяемыми, поэтому использование их в качестве аргументов по умолчанию нежелательно.

> Чтобы избежать ошибок, связанных с изменяемыми типами данных, используйте None в качестве значения аргумента по умолчанию и создавайте новый объект уже в теле функции, как это было сделано в примере выше.

# Способы передачи аргументов в python

В предыдущем юните мы создали функцию root, которую вы затем доработали, чтобы она автоматически считала квадратный корень. Вот она:


In [43]:
def root(value, n=2):
    result = value ** (1/n)
    return result

Эта функция написана так, что она возвращает результат в исходный код основного скрипта. Давайте сделаем её более «разговорчивой»: пусть функция печатает на экран, какие действия она выполнила, если мы укажем ей в качестве параметра слово verbose (от англ. «подробный», «многословный») со значением True. Чаще всего такой функционал будет излишним, поэтому зададим этому параметру значение по умолчанию False. Получим такую функцию:


In [44]:
def root(value, n=2, verbose=False):
    result = value ** (1/n)
    if verbose:
        # Аргументы в функции print,
        # перечисленные через запятую,
        # печатаются через пробел
        print('Root of power', n, 'from',
            value, 'equals', result)
    return result

Теперь давайте запустим новую функцию различными способами. Будем по порядку передавать в неё аргументы, которые функция ожидает получить (такие аргументы назовём порядковыми):

In [45]:
# Посчитаем корень из 25 с аргументами по умолчанию
print(root(25))
# Будет напечатано:
# 5.0
 
# Посчитаем кубический корень из 27 с verbose по умолчанию
print(root(27, 3))
# Будет напечатано:
# 3.0
 
#Посчитаем кубический корень из 27 с verbose=True
print(root(27, 3, True))
# Будет напечатано:
# Root of power 3 from 27 equals 3.0
# 3.0

5.0
3.0
Root of power 3 from 27 equals 3.0
3.0


Поведение функции выглядит совершенно ожидаемым: когда мы передали аргумент verbose=True, функция подробно напечатала результат выполнения.

А что делать, если мы хотим поменять только аргумент verbose, а степень корня оставить по умолчанию? 

Достаточно при передаче аргументов указать название того параметра, которому мы хотим присвоить значение, и передать это значение через = (равно) вот так:


In [46]:
print(root(25, verbose=True))
# Будет напечатано:
# Root of power 2 from 25 equals 5.0
# 5.0

Root of power 2 from 25 equals 5.0
5.0


Только что мы применили второй способ передачи аргументов — по ключевым словам. Такие аргументы называются именованными, поскольку мы можем написать имя аргумента и присвоить ему значение.

 На самом деле, мы сейчас использовали оба способа передачи аргументов: по порядку, так как функция root ожидает на первом месте получить аргумент value, и по ключевым словам. 

Теперь необходимо уточнить, в каком порядке необходимо сочетать порядковые и именованные аргументы.

Важно сначала передавать порядковые и только затем именованные аргументы.

Посмотрим, что будет, если передать в функцию сначала именованные аргументы и только потом value, который в данном случае является порядковым:


In [47]:
print(root(verbose=True, 25))
# Будет напечатано:
# SyntaxError: positional argument follows keyword argument

SyntaxError: positional argument follows keyword argument (2411210224.py, line 1)

Интерпретатор сообщил нам об ошибке. Её текст переводится так: «Синтаксическая ошибка: позиционный аргумент следует после именованного аргумента».

→ Запомните эту особенность передачи аргументов! 

Можно ли менять местами именованные аргументы?

Да, можно. Давайте так и сделаем:

In [48]:
# Подробно считаем корень степени 4 из 81
print(root(81, verbose=True, n=4))
# Будет напечатано:
# Root of power 4 from 81 equals 3.0
# 3.0

Root of power 4 from 81 equals 3.0
3.0


Как видите, порядок именованных аргументов никак не повлиял на работу программы. 

Есть небольшое исключение: можно записать порядковый аргумент после именованных, но для этого необходимо подать его в виде именованного — с помощью знака «равно». Вот так:


In [49]:
# Написали value=81 после именованных:
print(root(verbose=True, n=4, value=81))
# Будет напечатано:
# Root of power 4 from 81 equals 3.0
# 3.0

Root of power 4 from 81 equals 3.0
3.0


Как видите, при такой подаче аргументов ошибки не возникло.

# Аргументы по умолчанию в функции print

Функция print, с которой вы познакомились в самом начале курса по Python, на самом деле имеет множество дополнительных параметров, которыми бывает удобно пользоваться. Да, достаточно написать print("Hello world!") и сразу получить результат выполнения первой программы, но можно передать в функцию дополнительные именованные аргументы и изменить вывод. 

Удобно использовать следующие аргументы:
- sep - Этот аргумент задаёт строку, которая будет разделять позиционные аргументы, переданные на печать (по умолчанию это пробел).

- end - Этот аргумент задаёт символ, которым заканчивается печатаемая строка (по умолчанию это перенос на новую строку — символ \n).

Приведем пару примеров:

In [60]:
print(25, 125, 625)

print()

print(25, 125, 625, sep = ',')

print()

print("Shopping list:")
print("bread", "butter", "eggs")

print()

print("Shopping list:", end=' ')
print("bread", "butter", "eggs")

25 125 625

25,125,625

Shopping list:
bread butter eggs

Shopping list: bread butter eggs


# Обработка неизвестного заранее числа аргументов в PYTHON (*ARGS, *KWARGS)

Сейчас мы напишем функцию, которая принимает неизвестное заранее число аргументов для обработки, чтобы вы тоже научились создавать подобные функции.

Пусть функция mean (от англ. mean — «среднее значение») принимает на вход неограниченное количество чисел через запятую, а в конце возвращает их среднее значение. Она выглядит вот так:



In [61]:
# В массив args будут записаны все переданные
# порядковые аргументы
def mean(*args):
    # Среднее значение — это сумма всех значений,
    # делённая на число этих значений
    # Функция sum — встроенная, она возвращает
    # сумму чисел
    result = sum(args) / len(args)
    return result
 
# Передадим аргументы в функцию через запятую,
# чтобы посчитать их среднее
print(mean(5,4,4,3))

4.0


Только что вы увидели в аргументах запись *args. Здесь * — это не символ умножения, а оператор распаковки. Все порядковые аргументы будут записаны в переменную args. Args — это принятое в Python обозначение для порядковых аргументов, однако можно называть эту переменную любым другим подходящим образом. Например, конкретно для функции mean логично переименовать переменную args в numbers (от англ. «числа»):

In [62]:
def mean(*numbers):
    result = sum(numbers) / len(numbers)
    return result
 
print(mean(5,4,4,3))

4.0


Как видите, от переименования аргумента поведение функции не изменилось и ошибки не возникло. 

- Какой же тип данных скрывается за объектом numbers в теле функции?

Это кортеж (tuple):


In [63]:
def mean(*numbers):
    # С помощью встроенной функции isinstance
    # проверим, что numbers — это tuple
    print(isinstance(numbers, tuple))
    # Напечатаем содержимое объекта numbers
    print(numbers)
    result = sum(numbers) / len(numbers)
    return result
 
print(mean(5,4,4,3))

True
(5, 4, 4, 3)
4.0


Необязательно записывать все порядковые аргументы через * в args. Пусть функция mean_mark теперь печатает среднюю оценку для конкретного студента:

In [64]:
# В качестве первого аргумента принимаем фамилию
# студента, а затем уже его оценки через запятую
def mean_mark(name, *marks):
    result = sum(marks) / len(marks)
    # Не возвращаем результат, а печатаем его
    print(name+':', result)
 
mean_mark("Ivanov", 5, 5, 5, 4)
mean_mark("Petrov", 5, 3, 5, 4)

Ivanov: 4.75
Petrov: 4.25


Как видите, первый аргумент мы заняли под строковое значение — имя студента, а дальнейшие аргументы записали уже в marks.

После аргументов, записанных через *args, не могут идти другие порядковые аргументы:

In [65]:
# Поменяем местами *marks, name
def mean_mark_wrong(*marks, name):
    result = sum(marks) / len(marks)
    print(name+':', result)
 
mean_mark_wrong(5, 5, 5, 4, "Ivanov")

TypeError: mean_mark_wrong() missing 1 required keyword-only argument: 'name'

Возникла ошибка. Она переводится так: «Ошибка типов: в функции mean_mark_wrong отсутствует один обязательный аргумент 'name', который можно передать только через ключевые слова». Это означает, что аргумент name автоматически стал именованным. К тому же он является обязательным, поскольку не имеет значения по умолчанию.

Передать фамилию студента при таком порядке аргументов можно только вот так:

- Воспользуемся названием name, чтобы передать фамилию
- в этот аргумент
mean_mark_wrong(5, 5, 5, 4, name="Ivanov")
- Будет напечатано:
- Ivanov: 4.75
Теперь функция сработала без ошибок, однако аргумент name стал именованным.

> Необходимо запомнить, что все аргументы, следующие за конструкцией с оператором *, считаются именованными.

Оператор распаковки * можно использовать и для передачи значений из списка в функцию.

Рассмотрим эту возможность на примере функции print. У нас есть список:

In [67]:
langs = ['Python', 'SQL', 'Machine Learning', 'Statistics']

Напечатаем его разными способами. Самый простой:

In [68]:
print(langs)

['Python', 'SQL', 'Machine Learning', 'Statistics']


В данном случае список был напечатан с характерными для него квадратными скобками на границах и кавычками у строк. Так список превращается в строку по умолчанию.

Чтобы напечатать слова в списке langs через пробел, достаточно в функции print перед langs поставить * как оператор распаковки:

In [69]:
print(*langs)

Python SQL Machine Learning Statistics


Получается, что теперь функция print получила аргументы уже не в виде списка, а в виде перечисленных в функции нескольких порядковых аргументов.

> Оператор * удобно использовать для печати списков в более читабельном виде.

# KWARGS

Оказывается, можно передавать разное число именованных аргументов с помощью ещё одного оператора распаковки — двух звёздочек подряд (**).

Переменная kwargs — стандартное название для совокупности именованных аргументов в Python (сокр. от англ. keyword arguments — «именованные аргументы»).

Однако вместо kwargs можно использовать любое другое адекватное название переменной.

Напишем функцию schedule, которая будет печатать расписание по дням. Для начала узнаем, что будет храниться в переменной kwargs после использования оператора **:

In [70]:
# В переменную kwargs будут записаны все
# именованные аргументы
def schedule(**kwargs):
    # kwargs — это словарь, проверим это с помощью isinstance:
    print(isinstance(kwargs, dict))
    # Напечатаем объект kwargs
    print(kwargs)
 
schedule(monday='Python', tuesday='SQL', friday='ML')
# Будет напечатано:
# True
# {'monday': 'Python', 'tuesday': 'SQL', 'friday': 'ML'}

True
{'monday': 'Python', 'tuesday': 'SQL', 'friday': 'ML'}


Итак, kwargs является словарём, в котором ключами являются имена аргументов, а значениями — переданные аргументам значения.

Следует понимать, что всё-таки изначально в функцию передаётся не словарь: те объекты, которые могли бы быть ключами словаря (например, числа), при передаче в качестве названия аргумента вызовут ошибку:

In [73]:
schedule(monday='Python', tuesday='SQL', 5='ML')
# Будет напечатано:
# SyntaxError: keyword can't be an expression

SyntaxError: expression cannot contain assignment, perhaps you meant "=="? (1271462094.py, line 1)

Эта ошибка переводится так: «Синтаксическая ошибка: ключевое слово не может быть выражением». То есть ключевое слово должно удовлетворять тем же требованиям, что и переменные. 

Обращаться к значениям в kwargs можно так же, как и к значениям в обычном словаре. Расширим функцию schedule так, чтобы она более презентабельно печатала переданные значения:

In [74]:
def schedule(**kwargs):
    print("Week schedule:")
    for key in kwargs:
        print(key, kwargs[key], sep=' - ')
 
schedule(monday='Python', tuesday='SQL', friday='ML')

Week schedule:
monday - Python
tuesday - SQL
friday - ML


С помощью оператора ** можно передавать в функцию нескольких аргументов. Он используется для передачи именованных аргументов с помощью словаря, который в дальнейшем будет распакован. Кроме того, оператор ** является оператором распаковки, только для словарей. Ключами словаря выступают названия аргументов, а значениями — те значения аргументов, которые должны быть им присвоены.

Создадим словарь lessons (уроки)  и передадим его в функцию schedule, описанную выше:

In [75]:
lessons = {
    'Wednesday': 'Maths',
    'Thursday': 'SQL',
    'Friday': 'Statistics'
}
# Использовали оператор ** для распаковки словаря в набор
# значений именованных аргументов
schedule(**lessons)

Week schedule:
Wednesday - Maths
Thursday - SQL
Friday - Statistics


Принцип совместного использования *args и **kwargs логично вытекает из принципа перечисления аргументов: сначала — порядковые, затем — именованные. 

Вот функция, которая просто печатает все переданные в неё аргументы:

Вот функция, которая просто печатает все переданные в неё аргументы:

Важное замечание. Указывать аргументам значения по умолчанию можно только после того, как в функции перечислены все обязательные аргументы.

Поменяем в модифицированной функции root порядок аргументов:

def wrong_root(n=2, value):
    result = value ** (1/n)
    return result
Эта ошибка переводится так: «Синтаксическая ошибка: аргумент без значения по умолчанию следует после аргумента по умолчанию». Чтобы её избежать, необходимо указывать аргументы по умолчанию только после обязательных.

In [76]:
def print_args(*args, **kwargs):
    print(args)
    print(kwargs)
 
print_args(1,4,5,7, name='Ivanov', age=19, city='Moscow')

(1, 4, 5, 7)
{'name': 'Ivanov', 'age': 19, 'city': 'Moscow'}


Если поменять местами * и **, возникнет ошибка:

In [77]:
def print_args_wrong(**kwargs, *args):
    print(args)
    print(kwargs)
 
print_args_wrong(1,4,5,7, name='Ivanov', age=19, city='Moscow')

SyntaxError: invalid syntax (1181091138.py, line 1)

Также в функцию можно одновременно передавать и порядковые, и именованные аргументы через операторы распаковки.

Напечатаем три списка с помощью print: значения внутри списка разделим запятой, а сами списки — точкой с запятой. Для этого один раз создадим словарь how, в котором будут храниться значения, присваиваемые в дальнейшем аргументам sep и end.

Когда мы распакуем словарь в print, поведение функции будет аналогично тому, что мы бы получили при использовании функции print в виде:

In [79]:
list1 = [1,4,6,8]
list2 = [12, 45, 56, 190, 111]
list3 = ['Python', 'Functions']
 
# Один раз запишем в словарь параметры для печати
# через print: разделитель — запятая,
# окончание строки — точка с запятой
how = {'sep': ', ', 'end': '; '}
 
# Распаковываем и список, и словарь how
print(*list1, **how)
print(*list2, **how)
print(*list3, **how)

1, 4, 6, 8; 12, 45, 56, 190, 111; Python, Functions; 

# LAMBDA - Функция
-  относится к классу временных или анонимных функций Python. Она позволяет быстро создавать короткие однострочные функции, для которых нет необходимости пропимсывать целиком сигнатуру и оператор return.

Разберем пару примеров


In [81]:
def root(num):
    # Напоминание: в Python используется оператор **
    # для возведения числа в степень.
    # В математике возведение в степень ½ соответствует
    # вычислению квадратного корня.
    return num ** (1/2)

root(9)

3.0

А теперь запишем тоже самое с помощью **lambda-функции**:

In [83]:
root = lambda num: num**(1/2)

3.0

В переменной root теперь хранится функция, которая выполняет то же самое действие, которое мы описали с помощью «классической» функции через def.

Разберём особенности синтаксиса:

1. Сначала пишется слово lambda, которое сообщает компьютеру, что далее следует lambda-функция.

2. Затем пишутся названия аргументов, которые функция будет обрабатывать (если аргументов несколько, они перечисляются через запятую).

3. Далее ставится двоеточие, и на той же строке записывается значение, которое возвращает функция в зависимости от условий.

В данном случае функция возвращает результат возведения num в степень ½.

Обобщим нашу функцию на корень произвольной степени:

In [86]:
# Для получения корня произвольной степени от числа
# (например, корня степени 4) необходимо возвести исходное
# число в степень, равную единице, делённой на желаемую
# степень корня.
nth_root = lambda num, n: num**(1/n)

Чтобы воспользоваться этой функцией, можно написать, например:

In [87]:
print(nth_root(16, 4))

2.0


> В lambda-функции можно использовать и условия.

Например, напишем lambda-функцию, которая определяет, является ли число чётным (от англ. even — «чётный», odd — «нечётный») и возвращает слова «чётное» или «нечётное».

In [88]:
# Напоминание: оператор % используется для получения остатка
# от деления. Если остаток от деления на 2 равен 0, то
# число является чётным.
# Обратный слэш (\) используется в Python для того,
# чтобы перенести одну строку кода на следующую строку.
# Получается, что компьютер интерпретирует записанное ниже
# как одну строку.
is_even = lambda num: "even" if num % 2 == 0 \
    else "odd"

> Lambda-функции не лишены возможности применения операторов распаковки в своём коде.

Вот lambda-функция, которая принимает на вход переменное число порядковых аргументов и возвращает кортеж из них:

In [89]:
func_args = lambda *args: args

print(func_args(1,2,3,4,5))

(1, 2, 3, 4, 5)


> Наконец, можно принимать и args, и kwargs одновременно, как и в обычной функции.

Вот функция, принимающая полный набор аргументов:

In [90]:
full_func = lambda *args, **kwargs: (args, kwargs)
 
print(full_func(1,5,6,7,name='Ivan', age=25))

((1, 5, 6, 7), {'name': 'Ivan', 'age': 25})


- Однако не стоит использовать слишком сложные lambda-функции, так как это сделает код нечитаемым.

Посмотрите, как будет выглядеть функция, которая определяет, делится ли число на 2 или на 3 (не волнуйтесь, если не совсем понимаете, что в ней происходит — функция написана так специально, чтобы показать, что так лучше не делать):

In [91]:
scary_func = lambda num: 'divided by 2 and 3' \
    if num % 6 == 0 \
    else 'divided by 2' if num % 2 == 0 \
    else 'divided by 3' if num % 3 == 0 \
    else 'not divided by 2 nor 3'

Данная функция записана синтаксически верно, однако человеку, впервые читающему ваш код, разобраться в ней будет непросто. Да и вы сами спустя некоторое время с трудом сможете снова понять, что здесь происходит. 

- Для чего же тогда использовать lambda-функции, и почему они вообще называются анонимными, если мы постоянно присваивали им имена с помощью переменных?

> На самом деле использовать lambda-функцию удобно, если требуется передать простую функцию в качестве аргумента другой функции.

Мы с вами ещё познакомимся с некоторыми такими ситуациями, а пока что давайте применим lambda-функцию для сортировки списка. 

Создадим список имён: 

In [92]:
names = ['Ivan', 'Kim', 'German', 'Margarita', 'Simon']

Отсортируем список с помощью функции sort() и напечатаем его содержимое:

In [93]:
names.sort()
print(names)

['German', 'Ivan', 'Kim', 'Margarita', 'Simon']


А теперь представим, что мы хотим отсортировать слова в списке не по алфавиту, а по их длине. Для этого как раз будет удобно воспользоваться короткой lambda-функцией, которую мы передадим в качестве параметра key. Теперь функция сортировки уже будет сравнивать не слова в алфавитном порядке, а их длины:


In [94]:
names.sort(key=lambda name: len(name))
print(names)

['Kim', 'Ivan', 'Simon', 'German', 'Margarita']


Как видите, слова действительно оказались отсортированы по возрастанию количества букв в них. При этом функция lambda действительно оказалась одноразовой: мы не сможем снова сослаться на неё из кода скрипта.

- Что делать в случае, если мы хотим отсортировать список сразу по нескольким условиям: например, сначала по длине слова, а потом — для слов с одинаковой длиной — уже по алфавиту?

> Примечание. Метод sort() в случае строк сортирует сначала заглавные буквы, а затем — строчные.

Для этого достаточно написать функцию, которая возвращает кортеж из нескольких элементов, которые будут сравниваться. Вначале будут сравниваться первые элементы кортежа, при их совпадении — вторые, и так далее. 

Создадим список:

In [95]:
new_list = ['bbb', 'ababa','aaa', 'aaaaa',  'cc']

Отсортируем его по длине слов:

In [96]:
new_list.sort(key=lambda word: len(word))
print(new_list)

['cc', 'bbb', 'aaa', 'ababa', 'aaaaa']


Как видите, слова отсортированы по длине, но слова одинаковой длины при этом не отсортированы по алфавиту, потому что мы прямо не указывали функции sort() на это требование. 

Теперь отсортируем этот список с помощью lambda-функции, которая возвращает кортеж, где в качестве первого значения указана длина слова, а в качестве второго — само слово:

In [97]:
new_list.sort(key=lambda word: (len(word), word))
print(new_list)

['cc', 'aaa', 'bbb', 'aaaaa', 'ababa']


Теперь слова отсортированы и по длине, и по алфавиту.

> Обратите внимание, что в случае с lambda-функцией при возврате нескольких значений круглые скобки для обозначения кортежа обязательны!